### DISCUSSION TITLE
Why fastapi shutdown event doesn't dump my db ?

### DISCUSSION THREAD

**Discussion by joaopfg** [fastapi/fastapi#6299] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Description

I'm deploying an API server (with FastAPI) and a db server (with PostgreSQL) using docker-compose. I defined a shutdown event in the API server:

```
@app.on_event("shutdown")
def shutdown_event():
    subprocess.call(["bash", "dump_db.sh"])
```
    
The bash script `dump_db.sh` dumps the db in the filesystem of the container that deploys the API server. The dumped file is shared as a volume with the host. So, the host repository that started the containers should get updated dumps on every shutdown of the API server. However, it doesn't happen. Could it be some issue with the shutdown event ? If I understood, it should be triggered when I do `docker-compose down` , right ?

If I entry in the API server container filesystem (with `docker exec -it api bash`) and dump the db inside it, the host updates its dump file instantly as well, which proves the idea works.

### Example Code

Minimal reproducible example: 

https://drive.google.com/file/d/10GD677SDpFaV6ZFOVbc5IKQ8WlPGUJ1m/view?usp=sharing

### Steps to reproduce:

1) `docker-compose up`

2) Access the db server using any tool you want (pgAdmin, PyCharm professional etc), the login is `username` and the password is `password`. Modify one of the entries of the table `testmodel` and save your modification.

3) `docker-compose down`

4) Notice the file `populated_db.sql` of the host isn't updated with the change of step 2.

### Steps to show the problem seems to be with FastAPI shutdown event:

1) `docker-compose up`

2) Access the db server using any tool you want (pgAdmin, PyCharm professional etc), the login is `username` and the password is `password`. Modify one of the entries of the table `testmodel` and save your modification.

3) In a new terminal `docker exec -it api bash`

4) `bash dump_db.sh` in the container terminal that opened in step 3.

5) Notice the `populated_db.sql` of the host is updated.

### Operating System

Linux (Ubuntu 20.04)

### FastAPI Version

0.73.0

### Python Version

3.9.12

**Comment by mbroton** [comment]
Hi! Aside from the details, as far as I understand, you want to run bash script on FastAPI shutdown that saves some data in a text file (in directory that is mounted to host), which all should happen inside of docker container.

So, FastAPI role in this, should be to properly execute code in shutdown handler, after shutting down containers with `docker compose down` command.

I reproduced this simple case:
`main.py`:
```py3
from fastapi import FastAPI
import subprocess

app = FastAPI()


@app.on_event("shutdown")
async def on_shutdown():
    subprocess.call(["bash", "run.sh"])


@app.get("/test")
async def hello():
    return "Hello"
```
`run.sh`:
```bash
#!/bin/bash

echo "Works" > shared/status.txt
```
`Dockerfile`:
```
FROM python:3.9

WORKDIR /usr/src/fastapi_shutdown
RUN apt-get update
COPY requirements.txt ./
RUN pip3 install --upgrade pip && pip3 install -r requirements.txt
COPY . .
```
`docker-compose.yaml` (in which I mounted `shared` directory):
```
version: '3.7'

services:
  api:
    build: .
    command: uvicorn main:app
    ports:
      - 8000:8000
    volumes:
      - ./shared:/usr/src/fastapi_shutdown/shared
```
...and I can tell it's working fine. Demo:
![ubuntu1](https://user-images.githubusercontent.com/50829834/188478226-c1fe1ded-91e2-4fa1-aeac-19bcf4a16461.gif)

I would say that the problem is in your code, since I kinda reproduced your case and it works (aside database logic), but I don't know exactly what's the cause. (permissions? bash script?)

**Comment by joaopfg** [comment, marked as answer]
Turned out the reason was the bash script. I was using it to postpone the launch of the API server to when the db server was fully lauched. Detaching the API server from the db server (launching them with different docker-compose.yml) and putting the launch of the API server directly in the docker-compose.yml was a solution to the problem. Thanks !

However, I wonder if there is a way to launch the API server using the same docker-compose.yml of the db server to facilitate the deployment and still be able to dump the db on shutdown.

**Comment by mbroton** [comment]
If you want to run the API server after the database is launched and ready, you can use a healthcheck in `docker-compose` file:
```
services:
  api:
    # ...
    depends_on:
      db:
        condition: service_healthy
  db:
    # ...
    healthcheck:
      test: ["CMD-SHELL", "pg_isready", "-d", "postgres"]
      interval: 5s
      timeout: 5s
      retries: 3
```
This way you can still launch the API server directly from docker-compose and be sure that db is running.
But I'm not sure if it will solve your problem.

**Comment by joaopfg** [comment]
@mbroton Thanks, it solved all my problems !
